pkg/log/locallogger/locallogger.go

```diff
@@ -1,24 +1,248 @@
 package locallogger
 
 import (
+	"crypto/sha256"
+	"encoding/json"
 	"fmt"
 	"io"
+	"sort"
+	"strings"
+	"sync"
+	"time"
 
 	"github.com/go-kit/kit/log"
 	"gopkg.in/natefinch/lumberjack.v2"
 )
 
 const (
 	truncatedFormatString = "%s[TRUNCATED]"
+
+	// Deduplication configuration
+	defaultCacheExpiry   = 5 * time.Minute // How long to remember log entries
+	defaultMaxCacheSize  = 2000            // Maximum number of unique log entries to track
+	cleanupInterval      = 1 * time.Minute // How often to clean up expired entries
+	duplicateLogInterval = 1 * time.Minute // How long to wait before logging a duplicate again
 )
 
+// Fields to exclude when creating content hash for deduplication
+var excludedHashFields = map[string]bool{
+	"ts":              true, // go-kit timestamp
+	"time":            true, // slog timestamp
+	"caller":          true, // go-kit caller info
+	"source":          true, // slog source info
+	"original.time":   true,
+	"original.source": true,
+}
+
+// logEntry tracks information about seen log messages for deduplication
+type logEntry struct {
+	firstSeen  time.Time
+	lastSeen   time.Time
+	count      int
+	lastLogged time.Time
+}
+
+// dedupWriter wraps the actual writer and handles all deduplication logic for both log paths
+type dedupWriter struct {
+	actualWriter io.Writer
+
+	// Deduplication state
+	dedupCache   map[string]*logEntry
+	dedupMutex   sync.RWMutex
+	cacheExpiry  time.Duration
+	maxCacheSize int
+	lastCleanup  time.Time
+}
+
+// Write implements io.Writer and deduplicates JSON log lines before writing
+func (w *dedupWriter) Write(p []byte) (n int, err error) {
+	// Handle the case where we receive multiple log lines in one write
+	lines := strings.Split(strings.TrimSpace(string(p)), "\n")
+
+	for _, line := range lines {
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+
+		// Try to parse as JSON to extract fields for deduplication
+		var logData map[string]interface{}
+		if err := json.Unmarshal([]byte(line), &logData); err != nil {
```

Comment:
Needing to unmarshal and remarshal feels weird. 

I wonder if this would be any easier if you did this at the slog layer. I forget how wrappers there work. Maybe as handlers?